frontends/concrete-python/examples/sha1/sha1.md

```diff
@@ -0,0 +1,1702 @@
+        ans = fhe.zeros((32,))
+        cy = 0
+
+        for i in range(32):
+            t = x[i] + y[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        cy = 0
+
+        for i in range(32):
+            t = ans[i] + u[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        cy = 0
+
+        for i in range(32):
+            t = ans[i] + v[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        cy = 0
+
+        for i in range(32):
+            t = ans[i] + w[i] + cy
+            cy, tr = t // 2, t % 2
+            ans[i] = tr
+
+        return ans
+```
+
+It contains 7 functions which can be combined together:
+- `xor3` XORs three values together
+- `iftern` computes the ternary IF, i.e., c ? t : f
+- `maj` computes the majority function
+- `rotate30` rotates a 32-bit word by 30 positions
+- `rotate5` rotates a 32-bit word by 5 positions
+- `add2` adds two values together
+- `add5` adds five values together
+
+This module is then compiled. Remark we set `p_error=10**-8` to a very small value, to avoid
+errors in computations. Thanks to the Module feature, all these functions can be combined
+together, i.e., outputs of some of them can be reused as inputs of some other ones. It is very
+convenient to have a larger function, with some control-flow (conditions, branches, loops) done in
+the clear and operating with these small functions. In our case, it's done in
+`_process_encrypted_chunk_server_side` function.
+
+## Details of `_process_encrypted_chunk_server_side`
+
+`_process_encrypted_chunk_server_side` uses inputs which are encrypted, and returns encrypted
+values. Everything is a 32-bit word when done in the clear. Here, words will be represented as 32
+1-bit encrypted values, to simplify (and accelerate) the non-linear operations in SHA1.
+
+Then, we have the main loop of the compression function which is done in the clear with
+`for i in range(80):`. There, we also have the different conditions (`if 0 <= i <= 19:`,
+`elif 20 <= i <= 39` etc) which is done in the clear. Obviously, there is no security issue here,
+since the `i` index is not secret. Depending on `i`, we then select the right `fsplit_enc` and
+`ksplit`, which correspond to `f` and `k` in the clear implementation of the compression function.
+Here, we can see a first use of the different functions in the Module.
+
+Then, we continue with other functions in the module, to compute `arot5 = _left_rotate(a, 5)` and
+`arot5 + f + e + k + w[i]`. Finally, we update the different `a, b, c, d, e` values as in the clear
+implementation but with the encrypted forms.
+
+One can remark that we have compiled the different functions of the Module on inputset made with
+bits. Under the hood, Concrete will add a few programmable bootstrapping, to compute the right
+functions in FHE.
+
+## MLIR code
+
+Compiling with `show_mlir = True` allows to see the different MLIR implementations. Typically, with
```

Comment:
Not sure of why showing MLIR here, there are no explanation. And I feel that complicated the tutorial in a unnecessary way, the target for this tutorial don't care about the MLIR I guess.

Reply:
I agree!

Reply:
Removed it